packages/convex-helpers/server/zod4.ts

```diff
@@ -0,0 +1,642 @@
+        const added = await customInput(ctx, pick(allArgs, Object.keys(inputArgs)) as any, extra);
+        const finalCtx = { ...ctx, ...(added?.ctx ?? {}) };
+        const baseArgs = allArgs as Record<string, unknown>;
+        const addedArgs = (added?.args as Record<string, unknown>) ?? {};
+        const finalArgs = { ...baseArgs, ...addedArgs };
+        const ret = await handler(finalCtx, finalArgs);
+        if (returns && !fn.skipConvexValidation) {
+          let validated: any;
+          try {
+            validated = (returns as z.ZodTypeAny).parse(ret);
+          } catch (e) {
+            handleZodValidationError(e, "returns");
+          }
+          if (added?.onSuccess) await added.onSuccess({ ctx, args: allArgs, result: validated });
+          return toConvexJS(returns as z.ZodTypeAny, validated);
+        }
+        if (added?.onSuccess) await added.onSuccess({ ctx, args: allArgs, result: ret });
+        return ret;
+      },
+    });
+  };
+}
+
+export type ZCustomCtx<Builder> = Builder extends CustomBuilder<
+  any,
+  any,
+  infer CustomCtx,
+  any,
+  infer InputCtx,
+  any,
+  any
+>
+  ? Overwrite<InputCtx, CustomCtx>
+  : never;
+
+export function zCustomQuery<
+  CustomArgsValidator extends PropertyValidators,
+  CustomCtx extends Record<string, any>,
+  CustomMadeArgs extends Record<string, any>,
+  Visibility extends FunctionVisibility,
+  DataModel extends GenericDataModel,
+  ExtraArgs extends Record<string, any> = object,
+>(
+  query: QueryBuilder<DataModel, Visibility>,
+  customization: Customization<
+    GenericQueryCtx<DataModel>,
+    CustomArgsValidator,
+    CustomCtx,
+    CustomMadeArgs,
+    ExtraArgs
+  >,
+) {
+  return customFnBuilder(query, customization) as any;
+}
+
+export function zCustomMutation<
+  CustomArgsValidator extends PropertyValidators,
+  CustomCtx extends Record<string, any>,
+  CustomMadeArgs extends Record<string, any>,
+  Visibility extends FunctionVisibility,
+  DataModel extends GenericDataModel,
+  ExtraArgs extends Record<string, any> = object,
+>(
+  mutation: MutationBuilder<DataModel, Visibility>,
+  customization: Customization<
+    GenericMutationCtx<DataModel>,
+    CustomArgsValidator,
+    CustomCtx,
+    CustomMadeArgs,
+    ExtraArgs
+  >,
+) {
+  return customFnBuilder(mutation, customization) as any;
+}
+
+export function zCustomAction<
+  CustomArgsValidator extends PropertyValidators,
+  CustomCtx extends Record<string, any>,
+  CustomMadeArgs extends Record<string, any>,
+  Visibility extends FunctionVisibility,
+  DataModel extends GenericDataModel,
+  ExtraArgs extends Record<string, any> = object,
+>(
+  action: ActionBuilder<DataModel, Visibility>,
+  customization: Customization<
+    GenericActionCtx<DataModel>,
+    CustomArgsValidator,
+    CustomCtx,
+    CustomMadeArgs,
+    ExtraArgs
+  >,
+) {
+  return customFnBuilder(action, customization) as any;
+}
+
+export function zBrand<T extends z.ZodTypeAny, B extends string | number | symbol>(
```

Comment:
Can you please document this export?

```suggestion
/**
 * Add a brand to a zod validator. Used like `zBrand(z.string(), "MyBrand")`.
 * Compared to zod's `.brand`, this also brands the input type, so if you use
 * the branded validator as an argument to a function, the input type will also
 * be branded. The normal `.brand` only brands the output type, so only the type
 * returned by validation would be branded.
 *
 * @param validator A zod validator - generally a string, number, or bigint
 * @param brand A string, number, or symbol to brand the validator with
 * @returns A zod validator that brands both the input and output types.
 */
export function zBrand<T extends z.ZodTypeAny, B extends string | number | symbol>(
```